### PR TITLE
Handle duplicate keys in load_data task

### LINKED ISSUE #388: Handle duplicate keys in the TSV load_data task
## Problem

Some of our provider DAGs are successfully pulling data from the API but failing in the `load_data` task due to duplicate key errors. Example:

```
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "image_url_key"
DETAIL:  Key (url)=(https://upload.wikimedia.org/wikipedia/commons/4/44/Indiana_Presidential_Election_Results_2020.png) already exists.
```

This happens when a Provider API sends us multiple rows that have different `foreign_identifier`s but the same `url`.

## Description

There are a couple of cases we need to handle:

1. A provider script pulls multiple rows with different ids but the same `url`. No record with this url currently exists in the Catalog
    * **Insert the first row received, and discard the others**
2. A provider script pulls multiple rows with different ids but the same `url`. A record with this url currently exists in the Catalog, and its id matches one of the pulled records.
    * **Upsert the row with the matching id; discard all others**
3. A provider script pulls one or more rows with the same `url` as a record that exists in the Catalog, but none of them match the id of the Catalog's record.
    * **Discard all the new rows as they do not match existing information**

**COMPLETION CRITERIA**: When testing this work, we should re-trigger the following DAGs (which are experiencing this problem) and ensure that they run:
* Wikimedia Commons (image pull task)
* Cleveland Museum
* Metropolitan Museum

## Alternatives

* We could instead run a query to remove duplicates from the temp table before upserting into the main db. This would certainly be _much_ slower and dramatically increase the length of the ingestion workflow.
* We could also consider adding a caching step to the _pull_data_ task in the actual provider scripts, to discard duplicates as we find them. This would resolve case 1 as described above, but wouldn't be capable of handling the more complex cases. 

## Additional context

To our knowledge no checking for duplicates has ever existed. The duplication caused serious problems with ingestion that could not be detected until the data refresh, sometimes many hours into the workflow. The unique key index was added to the Catalog to help us identify the problem sooner.

## Implementation

- [x] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by stacimc**
## Fixes

Fixes #388

## Description

Currently we have several provider DAGs that successfully pull data, but then fail in the `load_data` step due to duplicate keys on the direct `audio_url` or `image_url`, which must be unique. This PR attempts to add handling to the load_data task to avoid attempts to upsert these duplicates.

This must handle two basic cases (and the combination of the two):
* A record exists in the Catalog already with a particular `url`. We attempt to upsert a new record which has a **different** foreign identifier but the **same** url. (This new record should be ignored).
* A provider script returns multiple records with different foreign identifiers but the same url. We attempt to upsert all of these at the same time. (Only the first should be inserted).

See [comment](https://github.com/WordPress/openverse-catalog/pull/395#issuecomment-1057561255) for additional edge case.

### Notes/TODO
* I'm not sure if the tests could be flaky -- they currently assume that if I insert first `A` and then `B` (duplicates) into a loading table and then upsert, `A` will be the one that gets upserted to the final table. This relies on `A` being the first row processed.
* Would love some feedback about whether this is viable from a performance perspective

### Alternatives considered

* It would be great if we could use something like `INSERT [...] ON CONFLICT (url)  DO NOTHING`, which would do exactly what we want -- catch the unique key violation and move on to the next row -- but we are already using the `ON CONFLICT` clause to [perform the upsert](https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/dags/common/loader/sql.py#L252), and it is not possible to support multiple conflict targets.
* We can add an `EXCEPTION` to the existing INSERT to simply [catch unique key violations](https://github.com/WordPress/openverse-catalog/commit/4e6b39e86ee7d30f83380ad876a4cfd308f9616b), but this applies to the _entire_ insert as a single transaction -- meaning if hit, we swallow the error but all inserts are rolled back and no rows are inserted at all.
* We could do an "old fashioned" upsert using `LOOP` and apply the `EXCEPTION` handling to each individual transaction, but this will be extremely costly and a huge drag on performance.

## Testing Instructions

Run unit tests (and verify they make sense 😄) 
`just test`

To test manually in your local environment, you'll need to run a provider DAG and make sure that it returns some duplicates. I made some modifications to the `jamendo.py` script, available in [this patch](https://gist.github.com/stacimc/aba8c8b183a54607ad61aa59de504cb3) that do the following:
* Reduce the LIMIT to 10 items per page in API response
* `break` after the first page returned by the API; this means that the `pull_data` task will complete very quickly so you do not need to manually set the task to `success` or wait a long time for it to finish
* Manually force a duplicate in the results by modifying one of the results to be a clone of the other, except for the foreign id.

So to test manually:
* Grab the patch https://gist.github.com/stacimc/aba8c8b183a54607ad61aa59de504cb3 and [apply](https://git-scm.com/docs/git-apply) it
* Run the `jamendo` provider DAG locally
* Verify that all tasks complete successfully
* Check the logs for the `pull_data` task and observe that 10 lines were written.
* Optionally, you can actually check your local MinIO console and find the TSV file that was created to visually confirm that you have a duplicate url 😄 
* Check the logs for `report_load_completion` and see that only 9 records were actually upserted

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by stacimc** [comment]
Unfortunately I thought of a remaining edge case here. Example:

Let's say the provider API returns two rows with different ids but the same url:
```
{ 'foreign_identifier': 'a', 'url': 'https://images.com/a/img.jpg', ... } // A
{ 'foreign_identifier': 'b', 'url': 'https://images.com/a/img.jpg', ... } // B
```
And the catalog happens to have an existing record with one of those foreign identifiers:
```
{ 'foreign_identifier': 'b', ... }
```
Ideally what we want is to _keep and upsert the new record that matches the existing foreign id_. So in this case we'd want to discard `A` and upsert `B`, in order to update data for the existing record.

The current implementation doesn't do this, though -- it just keeps the first record returned by the API with a given url.

It's hard to tell if this is a realistic real-world scenario 🤔

**Comment by stacimc** [comment]
Even without the edge case described in my earlier comment, the approach of using the _current contents of the Catalog_ as the source of truth has a few downsides. A scenario like this could arise:

* Ingestion scripts run and receive an image with foreign id `a` from the provider API. It is inserted into the Catalog
* On all future runs of the ingestion script, the same image is returned but with foreign id `b` and is consequently discarded. The image's popularity data/etc never gets updated as a result.

Wild card/maybe silly idea: is the `url`s must be unique, could we ignore foreign_ids during insertion and use the url as the "primary key"? The biggest problem would be that if the API returns multiple records with the same url in a single DAG run, whichever one is upserted last would overwrite the others 🤔 

It would be helpful to have a better understanding of when/why we're getting these duplicates from the API, so we could understand whether scenarios like these could ever happen.

**Review by AetherUnbound — COMMENTED** [review]
It took me a sec to wrap my head around the query changes, but I think that approach makes sense! The `url` column has an index but I don't believe `foreign_identifier` has an index of its own (it does have a unique index along with provider). I think our next step should be to test this out against a copy of our production catalog!

File: openverse_catalog/dags/common/loader/sql.py [hunk 1]
```diff
@@ -241,17 +241,27 @@ def upsert_records_to_db_table(
             column_inserts[column.db_name] = NULL
             column_conflict_values[column.db_name] = NULL
         else:
-            column_inserts[column.db_name] = column.upsert_name
+            # The direct_url is handled specially to ensure uniqueness and
+            # should not be added to the column_inserts.
+            if not column.db_name == col.DIRECT_URL.name:
+                column_inserts[column.db_name] = column.upsert_name
```
[the remark this answers is not in the document]
Comment (stacimc) [reply openverse_catalog/dags/common/loader/sql.py:247]: Oh, I think this is much clearer! ✨

**Comment by stacimc** [comment]
Looks like I've got a test failing now -- latest commit _shouldn't_ be the culprit. Taking a look 👀

File: openverse_catalog/dags/common/loader/sql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stacimc) [reply openverse_catalog/dags/common/loader/sql.py:248]: It turns out the test failure _was_ related to inverting the test logic -- it completely slipped my mind that while `direct_url` is handled differently in the `column_inserts` it does still need to be added to the `column_conflict_values`.

I think moving that above the conditional instead makes this clearer?
```
            column_conflict_values[column.db_name] = column.upsert_value

            # The direct_url is handled specially to ensure uniqueness and
            # should not be added to the column_inserts.
            if not column.db_name == col.DIRECT_URL.name:
                column_inserts[column.db_name] = column.upsert_name
```

The good news is that the unit tests are absolutely useful! 😅

[the remark this answers is not in the document]
Comment (AetherUnbound) [reply openverse_catalog/dags/common/loader/sql.py:248]: Ooooh shoot, I totally missed that other conditional, my bad!

**Review by AetherUnbound — APPROVED** [review]
Staci and I paired today to try and see what the performance hit on this would be by testing it against the full data in the catalog.

## TL;DR

Load times were _faster_ but we don't take that as an indication that they'll be consistently faster across the board because of the small sample size. At the very least they weren't significantly slower, which gives us confidence that this solution will work great :thumbs_up:

## TS;WM

Here's the average load times per workflow for the S3 -> postgres step:

```
airflow> select dag_id, SPLIT_PART(task_id, '_', 2) as media_type, avg(duration) as avg_duration, count(*) as run_count from task_instance where task_id like '%load_from_s3' and state='success' group by 1, 2;
+----------------------------+------------+--------------+-----------+
| dag_id                     | media_type | avg_duration | run_count |
|----------------------------+------------+--------------+-----------|
| cleveland_museum_workflow  | image      | 18.796675    | 2         |
| freesound_workflow         | audio      | 3.445939     | 2         |
| jamendo_workflow           | audio      | 4.3266495    | 2         |
| nypl_workflow              | image      | 1.831779     | 1         |
| phylopic_workflow          | image      | 0.63856525   | 4         |
| science_museum_workflow    | image      | 39.8647275   | 2         |
| staten_museum_workflow     | image      | 36.4282225   | 2         |
| stocksnap_workflow         | image      | 23.2559025   | 2         |
| wikimedia_commons_workflow | audio      | 0.653764     | 2         |
| wordpress_workflow         | image      | 1.845394     | 2         |
+----------------------------+------------+--------------+-----------+
```

The longer loads (`science_museum`, `staten_museum`, `stocksnap`) were all loading ~10<sup>4</sup> rows, while the others were loading considerably less than that (~10 to ~10<sup>3</sup>).

We ran the `clevelend_museum` and `staten_museum` workflows as part of this test. The Cleveland DAG had not successfully run before and reingested the entire data set every run, so it was a good candidate for determining if this fix would correctly de-duplicate records. The Staten DAG was one that ran relatively quickly and generated lots of records, so it was a good control DAG for seeing how much the load time for a typical DAG without duplicates would be affected.

These were the averages after running those two DAGs:

```
airflow> select dag_id, SPLIT_PART(task_id, '_', 2) as media_type, avg(duration) as avg_duration, count(*) as run_count from task_instance where task_id like '%load_from_s3' and state='success' group by 1, 2;
+----------------------------+------------+-------------------+-----------+
| dag_id                     | media_type | avg_duration      | run_count |
|----------------------------+------------+-------------------+-----------|
| cleveland_museum_workflow  | image      | 18.796675         | 2         |
| freesound_workflow         | audio      | 3.445939          | 2         |
| jamendo_workflow           | audio      | 4.3266495         | 2         |
| nypl_workflow              | image      | 1.831779          | 1         |
| phylopic_workflow          | image      | 0.63856525        | 4         |
| science_museum_workflow    | image      | 39.8647275        | 2         |
| staten_museum_workflow     | image      | 33.16817366666667 | 3         |
| stocksnap_workflow         | image      | 23.2559025        | 2         |
| wikimedia_commons_workflow | audio      | 0.653764          | 2         |
| wordpress_workflow         | image      | 1.845394          | 2         |
+----------------------------+------------+-------------------+-----------+
```

Note that Staten actually went _down_ while processing the same amount of records! The altered query is more work so we don't expect performance to improve consistently across the board, but it is notable that it wasn't drastically worse. Additionally, the Cleveland loads ran on a similar number of records as Staten/Science/Stocksnap, so it's encouraging that the load times were comparable.

Ultimately, our strategy still relies on accepting one record and dropping all others. We haven't investigated much into _why_ we're getting these duplicates from providers on items we thought were unique, but these cases aren't frequent enough to warrant trying to build the perfect solution. @stacimc is planning on adding (either in this PR or a follow up) a way for us to track how many duplicates we receive per ingestion, if any. This will help us gauge whether we need to put more effort into sorting those out down the road.

---

That to say, I think this looks great! :tada:

**Comment by zackkrida** [comment]
@obulat and @dhruvkb will either of you have a chance to review this PR this week?

**Review by obulat — APPROVED** [review]
I ran the local insert as described in the instructions, and it all worked well. Great to have all the tests so well documented! I can follow what is happening, and understand **why** it's happening.
One tiny suggestion: in the reporting log, in addition to how many records **were** upserted, it would probably be useful to say how many records were **NOT** upserted.
```
[2022-03-15, 09:00:30 +03] {reporting.py:29} INFO - 
*Provider*: `jamendo`
*Media Type*: `audio`
*Number of Records Upserted*: 9
*Duration of data pull task*: 52.71344560700527
```

**Comment by stacimc** [comment]
#411 created to add some more Slack messaging in a follow-up
